### PR TITLE
Add context-based post processing for linear features

### LINKED ISSUE #218: Easier saving of predictions as csv files
**Is your feature request related to a problem? Please describe.**
Currently adding predictions to the patch_df is a bit complex.
We should add a function that does this for us and saves as a csv file.

**Describe the solution you'd like**
-[ ] basically add a save_predictions method which adds predictions to existing patch_df and saves new updated predictions_df as csv file.

**Describe alternatives you've considered**
Clear docs

### PR DESCRIPTION
**Pull request by rwood-97**
### Summary

As per #339, this PR implements a post-processing script so that users can filter out false positives. 
This works for linear features or anything where you expect multiple patches to be clustered but solo patches would be false positive.
It also adds a `save_predictions()` method the classifier to make sure predictions and confidence scores are saved in format expected for post-processing.

Fixes #218 
Addresses #339 

### Checklist before assigning a reviewer (update as needed)

- [x] Allow users to pick lowest conf for which to change label
- [x] Check for edge cases - overlapping patches and non-square patches
- [x] Self-review code
- [x] Ensure submission passes current tests
- [x] Add tests
- [ ] Update relevant docs

### Reviewer checklist

Please add anything you want reviewers to specifically focus/comment on.

- [x] Everything looks ok?

### CODE REVIEW

**Comment by rwood-97** [comment]
To run, first run inference on some patches and save outputs by calling `my_classifier.save_predictions(set_name="infer")` with your chosen dataset. 

Then:

``` python
import pandas as pd

from mapreader.process.post_process import PatchDataFrame

df = pd.read_csv("./predictions_patch_df.csv", index_col=0)

labels_map = {
    0: "no",
    1: "railspace",
    2: "building",
    3: "railspace+building"
}

patches = PatchDataFrame(df, labels_map=labels_map)
patches.get_context(labels=["railspace", "railspace+building",])
patches.update_preds(remap={"railspace": "no", "railspace+building": "building"}, conf=0.8)
```

This will select all railspace/railspace+building patches, get their context, then update predictions for patches with no surrounding railspace and confidence score of less than 0.8. 

Can also set `remap` to be new labels completely (e.g. `{"railspace": "check me", "railspace+building": "check me"}`).

File: mapreader/process/post_process.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+        context_df = self[self["image_id"].isin(context_list)]
+        # drop central patch from context
+        context_df.drop(index=ix, inplace=True)
+
+        # reverse the labels_map dict
+        label_index_dict = {v: k for k, v in self.labels_map.items()}
+
+        prefix = "" if inplace else "new_"
+        if (not any(context_df["predicted_label"].isin(labels))) & (
```
Comment (edwardchalstrey1) [inline mapreader/process/post_process.py:174]: I *think* that this line is the filtering rule? If so, might it make sense to separate it as it's own function and give it a name. The name could be `filter_rule_1` with a docstring that explains the logic in plain English like "If patch has label X but no surrounding patches have label X, remove label X from patch"

What you could do then is have an arg for `update_preds` called `filter_func`. You could pass in the filter function when it's called and the filter function would be called on this line. This would mean you could substitute in a different filter function.

So all you have to do to try out a different filter rule is write another small function and call update_preds like `patches.update_preds(filter_func=filter_rule_2 ... other args)`

File: mapreader/process/post_process.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            self.at[ix, "conf"] < conf
+        ):
+            self.at[ix, f"{prefix}predicted_label"] = remap[
```
Comment (edwardchalstrey1) [inline mapreader/process/post_process.py:177]: After this change is made to the `patches` object, would you then save another csv called like "predictions_patch_df_filtered.csv" or something?

Or if you implemented my suggestion above maybe "predictions_patch_df_filter_rule_1.csv"

File: mapreader/process/post_process.py (hunk 1 above)
Comment (rwood-97) [reply mapreader/process/post_process.py:174]: I'm going to make a follow up issue for this as an extra feautre vs within this PR.

**Comment by rwood-97** [comment]
See here for stats on post-processing https://github.com/Living-with-machines/railspace/issues/14

File: docs/source/User-guide/Post-process.rst [hunk 3]
```diff
@@ -1,4 +1,92 @@
 Post-process
 =============
 
-TBC
+MapReader post-processing's sub-package currently contains one method for post-processing the predictions from your model based on the idea that features such as railways, roads, coastlines, etc. are continuous and so patches with these labels should be found near to other patches also with these labels.
+
+For example, if a patch is predicted to be a railspace, but is surrounded by patches predicted to be non-railspace, then it is likely that the railspace patch is a false positive.
```
Comment (edwardchalstrey1) [inline docs/source/User-guide/Post-process.rst:6]: I guess you could be even more explicit and say: "The current method checks whether any of the 8 surrounding patches have the same label as a given patch (e.g. railspace), and if not, assumes this to be a false positive".

Perhaps could also mention: "Future releases may add functionality to create custom filter rules for your use case"

**Review by edwardchalstrey1 — APPROVED** [review]
One comment, but otherwise LGTM
